add imread

thewoz opened the pull request:
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [X] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

Hello everyone,
I created this new version of the imread function and I think it can be very useful in several cases.
It is actually passed to it object on which you want to upload the image.
The advantages can be different like in case one needs to open several large images all the same in sequence.
one can use the same pointer and the system would not allocate memory each time

dkurt commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -227,6 +227,10 @@ Currently, the following file formats are supported:
 CV_EXPORTS_W Mat imread( const String& filename, int flags = IMREAD_COLOR );
 
+
+CV_EXPORTS_W Mat imread( const String& filename, Mat & img, int flags = IMREAD_COLOR );
```
I think proper docstring may resolve some confusion about the proposed feature.

dkurt commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -639,6 +639,27 @@ Mat imread( const String& filename, int flags )
+ *  This function merely calls the actual implementation above and returns itself.
+ *
+ * @param[in] filename File to load
+ * @param[in] img
```
add a description

dkurt commented:
First, I was confused why there is a need to have `imread` with pre-allocated memory while `VideoCapture` can do something like that (for example, read a sequence of the same size frames from a disc). Then, I though that it might be useful.

thewoz commented:
Hi @dkurt thank you very much
I like to improve the documentation but my English is really bad

dkurt commented:
@thewoz, you may refer to current imread by `@ref imread(const String&, int)` but add a note that `img` param can be pre-allocated and memory is reused if sizes match.

dkurt commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -639,6 +639,27 @@ Mat imread( const String& filename, int flags )
+ * @param[in] img object in which the image will be loaded
+ * @param[in] flags Flags you wish to set.
+*/
+Mat imread( const String& filename, Mat & img, int flags )
```
Can you please add a test to `opencv/modules/imgcodecs/test/test_png.cpp`?

dkurt replied:
Please try this method with a ROI:
```cpp
Mat src = Mat::zeros(1000, 1000, CV_8UC3);
imread("image_640x480.png", src.rowRange(0, 480).colRange(0, 640));
```
Expected that final image it located in left-top corner and zero (black) pixels are not corrupted.

Adding a test like this also would be nice.

dkurt commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -227,6 +227,19 @@ Currently, the following file formats are supported:
+@note
+The image passing trought the img parameter can be pre-allocated. The memory is reused if the shape and the type match with the load image.
+ */
+CV_EXPORTS_W Mat imread( const String& filename, Mat & img, int flags = IMREAD_COLOR );
```
As this method marked as `CV_EXPORTS_W` and wrapped to Python, please add a test to /home/dkurt/opencv/modules/python/test. This is useful for users to see how it works and control regressions.

thewoz commented:
Hello everyone, 
I have been trying to do what you have suggested to do.
About @opencv-alalek 
On point (1) I tried to do it but you then give me error on conversion.
On point (2) I did not understand what you mean.
On point (3) I removed the return from the function.
About @dkurt 
I added the test in C++ . This is my first time doing this and I am not sure if I did it right. 
On adding the python test unfortunately I wouldn't really know where to put my hands.

dkurt commented on modules/imgcodecs/test/test_png.cpp:
```diff
@@ -7,6 +7,19 @@ namespace opencv_test { namespace {
+  Mat ori = imread("../cv/shared/lena.png");
+  Mat diff;
+  absdiff(ori, subImg, diff);
+  EXPECT_EQ(0, sum(diff)[0]);
```
Replace to `EXPECT_EQ(cv::norm(ori, subImg, NORM_INF), 0)`

thewoz replied:
done

dkurt commented on modules/imgcodecs/test/test_png.cpp:
```diff
@@ -7,6 +7,19 @@ namespace opencv_test { namespace {
+{
+  Mat img = Mat::zeros(1000, 1000, CV_8UC3);
+  Mat subImg = img.rowRange(0, 512).colRange(0, 512);
+  imread("../cv/shared/lena.png", subImg);
```
use `findData`

thewoz commented:
I hope I understood what you meant

opencv-alalek commented on modules/imgcodecs/test/test_png.cpp:
```diff
@@ -7,6 +7,17 @@ namespace opencv_test { namespace {
+TEST(imRead, passing_Mat)
+{
+  Mat img = Mat::zeros(1000, 1000, CV_8UC3);
+  Mat subImg = img.rowRange(0, 512).colRange(0, 512);
```
> `rowRange(0, 512)`

`rowRange(64, 512)`

> `colRange(0, 512);`

`colRange(128, 512);`

thewoz replied:
sorry I don't understand what you are asking me to do here. Why do you want to copy it to another interval? this way it would then fail the test

dkurt replied:
In that case test will check multiple strides as ROI will be an internal. Just a more strict test.

thewoz replied:
ok right

thewoz replied:
ok

opencv-alalek commented on modules/imgcodecs/test/test_png.cpp:
```diff
@@ -7,6 +7,17 @@ namespace opencv_test { namespace {
+  Mat subImg = img.rowRange(0, 512).colRange(0, 512);
+  imread(samples::findFile("lena.png"), subImg);
+  Mat ori = imread(samples::findFile("lena.png"));
+  EXPECT_EQ(cv::norm(ori, subImg, NORM_INF), 0);
```
Signature for `_EQ` is `ASSERT_EQ(expected_reference, actual_result);`.
Correct order of parameters are required to emit valid error messages.

Reference: https://github.com/opencv/opencv/blob/4.0.0/modules/ts/include/opencv2/ts/ts_gtest.h#L8196-L8200

```
GTEST_API_ AssertionResult EqFailure(const char* expected_expression,
                                     const char* actual_expression,
                                     const std::string& expected_value,
                                     const std::string& actual_value,
                                     bool ignoring_case);
```

opencv-alalek commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -227,6 +227,19 @@ Currently, the following file formats are supported:
+@param img object in which the image will be loaded.
+@param flags Flag that can take values of cv::ImreadModes
+@note
+The image passing trought the img parameter can be pre-allocated. The memory is reused if the shape and the type match with the load image.
```
> trought

typo

opencv-alalek commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -227,6 +227,19 @@ Currently, the following file formats are supported:
+
+This is an overloaded member function, provided for convenience. It differs from the above function only in what argument(s) it accepts and the return value.
+@param filename Name of file to be loaded.
+@param img object in which the image will be loaded.
```
img => dst

thewoz replied:
fixed

opencv-alalek commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -227,6 +227,19 @@ Currently, the following file formats are supported:
+@note
+The image passing trought the img parameter can be pre-allocated. The memory is reused if the shape and the type match with the load image.
+ */
+CV_EXPORTS_W void imread( const String& filename, Mat & img, int flags = IMREAD_COLOR );
```
Why `OutputArray dst` doesn't work? What is the error message?

thewoz replied:
/Users/thewoz/Documents/GitHub/opencv/modules/imgcodecs/src/loadsave.cpp:657:5: error: no matching function for call to 'imread_'
    imread_( filename, flags, dst );
    ^~~~~~~
/Users/thewoz/Documents/GitHub/opencv/modules/imgcodecs/src/loadsave.cpp:395:1: note: candidate function not viable: no known conversion from 'const _OutputArray' to 'Mat &' for 3rd argument
imread_( const String& filename, int flags, Mat& mat )
^
1 error generated.

dkurt replied:
Use `.getMat` method of `OutputArray`

opencv-alalek commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -227,6 +227,19 @@ Currently, the following file formats are supported:
+@note
+The image passing trought the img parameter can be pre-allocated. The memory is reused if the shape and the type match with the load image.
+ */
+CV_EXPORTS_W void imread( const String& filename, Mat & img, int flags = IMREAD_COLOR );
```
> `CV_EXPORTS_W void imread`

There are 2 options to avoid conflicts and ambiguity:
1. imread => imreadToBuffer
2. CV_EXPORTS_W => CV_EXPORTS_AS(imreadToBuffer)

dkurt replied:
I don't have conflicts in Python binding using just `OuputArray`:
```python
import cv2 as cv

img = cv.imread("img1.png")
cv.imread("img2.png, img)
```

thewoz replied:
it's all the same to me. how do we want to proceed?

opencv-alalek commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -639,6 +639,25 @@ Mat imread( const String& filename, int flags )
+ * @param[in] filename File to load
+ * @param[in] img object in which the image will be loaded
+ * @param[in] flags Flags you wish to set.
+*/
```
Docstrings in .cpp files are usually not used.

thewoz replied:
I put Docstrings in as I saw that the other imread functions had them. if you prefer I can remove them

dkurt replied:
Yes, let's avoid docstring in the implementation

opencv-alalek reviewed (commented):
Need to add Python tests (modules/imgproc/misc/python/test/test_imread.py).
See https://github.com/opencv/opencv/blob/4.x/modules/dnn/misc/python/test/test_dnn.py as reference.

Run Python test from build directory:

```
OPENCV_PYTHON_DEBUG=1 OPENCV_PYTEST_FILTER=test_imread* ./setup_vars.sh python3 ${OPENCV_SRC_DIR}/modules/python/test/test.py -v --repo ${OPENCV_SRC_DIR}
```

dkurt commented on modules/imgcodecs/test/test_png.cpp:
```diff
@@ -7,6 +7,17 @@ namespace opencv_test { namespace {
+{
+  Mat img = Mat::zeros(1000, 1000, CV_8UC3);
+  Mat subImg = img.rowRange(0, 512).colRange(0, 512);
+  imread(samples::findFile("lena.png"), subImg);
```
```suggestion
  imread(samples::findFile("cv/shared/lena.png"), subImg);
```

thewoz commented:
I tried to write the test in python
I am not sure if it is correct as I do not normally program in python

dkurt commented on modules/imgcodecs/test/test_png.cpp:
```diff
@@ -7,6 +7,19 @@ namespace opencv_test { namespace {
+  EXPECT_EQ(0, norm(sum(img.rowRange(0, 1000).colRange(0, 128))));
+  EXPECT_EQ(0, norm(sum(img.rowRange(0, 1000).colRange(640, 1000))));
+  EXPECT_EQ(0, norm(sum(img.rowRange(0, 64).colRange(128, 640))));
+  EXPECT_EQ(0, norm(sum(img.rowRange(576, 1000).colRange(128, 640))));
```
```
cv::norm
```
everywhere

dkurt commented on modules/python/test/test_imread.py:
```diff
@@ -0,0 +1,18 @@
+assert cv2.norm(np.sum(img[0:1000, 0:128])) == 0
+assert cv2.norm(np.sum(img[0:1000, 640:1000])) == 0
+assert cv2.norm(np.sum(img[0:64, 128:640])) == 0
+assert cv2.norm(np.sum(img[576:1000, 128:640])) == 0
```
Check https://github.com/opencv/opencv/blob/e7065b50547b19552ddda6f8e8c3abfea188b684/modules/python/test/test_dft.py as a reference. Use `NewOpenCVTests` class

thewoz replied:
ok

dkurt commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -639,6 +639,15 @@ Mat imread( const String& filename, int flags )
+
+    /// load the data
+    imread_( filename, flags, dst.getMatRef() );
+
```
extra whitespace

dkurt replied:
still relevant about empty line

dkurt commented on modules/imgcodecs/test/test_png.cpp:
```diff
@@ -7,6 +7,19 @@ namespace opencv_test { namespace {
+  Mat img = Mat::zeros(1000, 1000, CV_8UC3);
+  Mat subImg = img.rowRange(64, 576).colRange(128, 640);
+  imread(samples::findFile("../cv/shared/lena.png"), subImg);
+  Mat ori = imread(samples::findFile("../cv/shared/lena.png"));
```
@thewoz, the path is still invalid

thewoz replied:
this is what I get with  ../cv/shared/lena.png 

C++ exception with description "OpenCV(4.8.0-dev) /Users/thewoz/Github/opencv/modules/core/src/utils/samples.cpp:64: error: (-2:Unspecified error) OpenCV samples: Can't find required data file: ../cv/shared/lena.png in function 'findFile' " thrown in the test body.

this is what I get with  /cv/shared/lena.png 

C++ exception with description "OpenCV(4.8.0-dev) /Users/thewoz/Github/opencv/modules/core/src/utils/samples.cpp:64: error: (-2:Unspecified error) OpenCV samples: Can't find required data file: /cv/shared/lena.png in function 'findFile' " thrown in the test body.


instead this is working

 const string root = cvtest::TS::ptr()->get_data_path();
 const string imgName = root + "../cv/shared/lena.png";

opencv-alalek replied:
> this is what I get with /cv/shared/lena.png

Just remove leading slash.

thewoz replied:
Hi, I tray also in this way.. "cv/shared/lena.png"  but I got  the same result.

C++ exception with description "OpenCV(4.8.0-dev) /Users/thewoz/Dropbox/Research/Github/opencv/modules/core/src/utils/samples.cpp:64: error: (-2:Unspecified error) OpenCV samples: Can't find required data file: cv/shared/lena.png in function 'findFile'
" thrown in the test body.

dkurt commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -227,6 +227,19 @@ Currently, the following file formats are supported:
+ */
+CV_EXPORTS_W void imread( const String& filename, OutputArray dst, int flags = IMREAD_COLOR );
+
+
```
Keep single empty line before and after

dkurt commented on modules/imgcodecs/test/test_png.cpp:
```diff
@@ -7,6 +7,19 @@ namespace opencv_test { namespace {
+  EXPECT_EQ(0, cv::norm(sum(img.rowRange(0, 1000).colRange(0, 128))));
+  EXPECT_EQ(0, cv::norm(sum(img.rowRange(0, 1000).colRange(640, 1000))));
+  EXPECT_EQ(0, cv::norm(sum(img.rowRange(0, 64).colRange(128, 640))));
+  EXPECT_EQ(0, cv::norm(sum(img.rowRange(576, 1000).colRange(128, 640))));
```
Use NORM_INF everywhere

dkurt replied:
```
EXPECT_EQ(0, countNonZero(img.rowRange(0, 1000).colRange(0, 128)));
```
and the same for others

dkurt commented on modules/imgcodecs/test/test_png.cpp:
```diff
@@ -7,6 +7,19 @@ namespace opencv_test { namespace {
+  Mat subImg = img.rowRange(64, 576).colRange(128, 640);
+  imread(samples::findFile("/cv/shared/lena.png"), subImg);
+  Mat ori = imread(samples::findFile("/cv/shared/lena.png"));
+  EXPECT_EQ(0, cv::norm(ori, subImg, NORM_INF));
```
You have to compare with img ROI again. `subImg` can be reallocated and test won't catch this.

```cpp
Mat img = Mat::zeros(1000, 1000, CV_8UC3);
Mat subImg = img.rowRange(64, 576).colRange(128, 640);
imread(samples::findFile("/cv/shared/lena.png"), subImg);

subImg = img.rowRange(64, 576).colRange(128, 640);
Mat ori = imread(samples::findFile("cv/shared/lena.png"));
EXPECT_EQ(0, cv::norm(ori, subImg, NORM_INF));
```

dkurt replied:
@thewoz , I refered this comment for your question in https://github.com/opencv/opencv/pull/24415#issuecomment-1827506930 on how to use `findFile` and what is a correct path to test image.

thewoz replied:
ah ok sorry

thewoz replied:
Hi @dkurt,

I've tried three different ways, and I don't understand what the problem is.

NOTE: the $OPENCV_TEST_DATA_PATH is set in the correct way

**- [1] samples::findFile("samples/data/lena.jpg")**

I got this warning but eventually the image is loaded correctly

`[ WARN:0@0.435] global samples.cpp:61 findFile cv::samples::findFile('samples/data/lena.jpg') => '/Users/thewoz/Dropbox/Research/Github/opencv1/build/..//samples/data/lena.jpg'`

**- [2] findDataFile("cv/shared/lena.png")**

I got this error:

`unknown file: Failure
C++ exception with description "OpenCV(4.8.0-dev) /Users/thewoz/Dropbox/Research/Github/opencv1/modules/ts/src/ts.cpp:1071: error: (-2:Unspecified error) OpenCV tests: Can't find required data file: cv/shared/lena.png in function 'findData'
" thrown in the test body.`

this I do not understand why it does not work as it is called in other files:

modules/dnn/test/test_caffe_importer.cpp:    Mat img = imread(findDataFile("cv/shared/lena.png"));
modules/dnn/test/test_tflite_importer.cpp:    Mat input = imread(findDataFile("cv/shared/lena.png"));
modules/dnn/test/test_tflite_importer.cpp:    Mat input = imread(findDataFile("cv/shared/lena.png"));
modules/dnn/test/test_torch_importer.cpp:    Mat sample = imread(findDataFile("cv/shared/lena.png"));

**- [3] samples::findFile("cv/shared/lena.png")**

I got this:

`[ WARN:0@0.437] global samples.cpp:61 findFile cv::samples::findFile('cv/shared/lena.png') => ''
unknown file: Failure
C++ exception with description "OpenCV(4.8.0-dev) /Users/thewoz/Dropbox/Research/Github/opencv1/modules/core/src/utils/samples.cpp:64: error: (-2:Unspecified error) OpenCV samples: Can't find required data file: cv/shared/lena.png in function 'findFile'
" thrown in the test body.`

I could not find any other examples of calls to "cv::samples::findFile('cv/xxxxx') in the code.
I understand that cv::samples::findFile() cerca nel OPENCV_TEST_DATA_PATH

How can I solve it?
sorry so much for the trouble

dkurt replied:
C++ Test passed on CI: https://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/105434

now just make Python happy (replace cv2 usage to cv)

thewoz replied:
Sorry but Python is not my favorite language :)
This is how I set the environment variable:
export OPENCV_TEST_DATA_PATH=/Users/thewoz/Github/opencv_extra/testdata/
It's correct?

dkurt commented on modules/python/test/test_imread.py:
```diff
@@ -0,0 +1,32 @@
+        img = np.zeros((1000, 1000, 3), dtype=np.uint8)
+        subImg = img[64:576, 128:640]
+        subImg = cv2.imread("/cv/shared/lena.png", subImg)
+        ori = cv2.imread("/cv/shared/lena.png")
```
invalid paths

dkurt commented on modules/python/test/test_imread.py:
```diff
@@ -0,0 +1,32 @@
+ 
+        img = np.zeros((1000, 1000, 3), dtype=np.uint8)
+        subImg = img[64:576, 128:640]
+        subImg = cv2.imread("/cv/shared/lena.png", subImg)
```
Use `self.find_file`

dkurt commented on modules/python/test/test_imread.py:
```diff
@@ -0,0 +1,32 @@
+    def test_dft(self):
+ 
+        img = np.zeros((1000, 1000, 3), dtype=np.uint8)
+        subImg = img[64:576, 128:640]
```
This is 513x513

thewoz commented:
It remains to check the python test and what name to give to the function. 
So far I have not been able in the test to load the lena image using findFile()
Anyway, I changed two lines of code in the imread_ function. 
I put a check so that in this new version of the imread function the memory is ever allocated. 
In fact if that happened the example written in test would fail.

thewoz commented:
Hello everyone,
I wanted to know if there was anything I could do.
I'm trying to make a point.
I keep having problems with the findFile() function.

I have tried with:
../cv/shared/lena.png
/cv/shared/lena.png
cv/shared/lena.png

keeps giving me as error:
"OpenCV samples: Can't find required data file: cv/shared/lena.png in function 'findFile'"

what can i do?

as for python's test what do you think?

dkurt commented:
@thewoz, https://github.com/opencv/opencv/pull/24415/files#r1386723679

dkurt commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -633,12 +638,20 @@ Mat imread( const String& filename, int flags )
+    CV_TRACE_FUNCTION();
+
+    /// load the data
+    imread_(filename, flags, dst.getMatRef(), false);
```
I

opencv-alalek replied:
> `.getMatRef`

It should not be used in general because it restricts parameter type to `Mat` only.

dkurt commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -464,8 +464,13 @@ imread_( const String& filename, int flags, Mat& mat )
+    else 
+    {
+      CV_Assert(size == mat.size());
+      CV_Assert(type == mat.type());
```
I don't see any problem on reallocation of `mat` in case of size/type mismatch. Other OpenCV methods will behave in a same way.

thewoz replied:
Hi @dkurt,
In principle I agree with you that there would be no problem with reallocation.
But the idea of this version of imread is that it uses a given area of memory (perhaps shared).
So in this case I expect that if imread can't use the memory I'm passing, it will give me an error.
Internal reallocation without warning can lead to strange behaviour. 

For example in the code below:

```
Mat img = Mat::zeros(1000, 1000, CV_8UC3);
Mat subImg = img.rowRange(64, 576).colRange(128, 640);
imread(samples::findFile("cv/shared/lena.png"), subImg);
```

the _subImg_ would be reallocated while _img_ remain an empty image.

dkurt commented on modules/imgcodecs/test/test_png.cpp:
```diff
@@ -51,7 +66,7 @@ TEST(Imgcodecs_Png, regression_ImreadVSCvtColor)
 
 // Test OpenCV issue 3075 is solved
-TEST(Imgcodecs_Png, read_color_palette_with_alpha)
+TEST(, read_color_palette_with_alpha)
```
Please revert

dkurt commented on modules/python/test/test_imread.py:
```diff
@@ -0,0 +1,31 @@
+ 
+        img = np.zeros((1000, 1000), dtype=np.uint8)
+        subImg = img[64:576, 128:640]
+        subImg = cv2.imread(self.find_file("/cv/shared/lena.png"), subImg, cv2.IMREAD_GRAYSCALE)
+        ori = cv2.imread(self.find_file("/cv/shared/lena.png"), cv2.IMREAD_GRAYSCALE)
```
```suggestion
        subImg = cv2.imread(self.find_file("cv/shared/lena.png"), subImg, cv2.IMREAD_GRAYSCALE)
        ori = cv2.imread(self.find_file("cv/shared/lena.png"), cv2.IMREAD_GRAYSCALE)
```

dkurt replied:
```suggestion
        subImg = cv2.imread(self.find_file("cv/shared/lena.png"), subImg, cv2.IMREAD_GRAYSCALE)
        ori = cv2.imread(self.find_file("cv/shared/lena.png"), cv2.IMREAD_GRAYSCALE)
```

dkurt commented on modules/python/test/test_imread.py:
```diff
@@ -0,0 +1,31 @@
+ 
+        img = np.zeros((1000, 1000), dtype=np.uint8)
+        subImg = img[64:576, 128:640]
+        subImg = cv2.imread(self.find_file("/cv/shared/lena.png"), subImg, cv2.IMREAD_GRAYSCALE)
```
There is no `cv2` imported - use `cv.IMREAD_GRAYSCALE`

dkurt commented on modules/imgcodecs/test/test_png.cpp:
```diff
@@ -7,6 +7,21 @@ namespace opencv_test { namespace {
+  const string root = cvtest::TS::ptr()->get_data_path();
+  const string imgName = root + "../cv/shared/lena.png";
+  Mat img = Mat::zeros(1000, 1000, CV_8UC1);
+  Mat subImg = img.rowRange(64, 576).colRange(128, 640);
+  imread(imgName, subImg, IMREAD_GRAYSCALE);
+  Mat ori = imread(imgName, IMREAD_GRAYSCALE);
+  EXPECT_EQ(0, cv::norm(ori, subImg, NORM_INF));
```
Get ROI again before the check:

```suggestion
  Mat subImg = img.rowRange(64, 576).colRange(128, 640);
  imread(imgName, subImg, IMREAD_GRAYSCALE);
  
  Mat ori = imread(imgName, IMREAD_GRAYSCALE);
  subImg = img.rowRange(64, 576).colRange(128, 640);
  EXPECT_EQ(0, cv::norm(ori, subImg, NORM_INF));
```

dkurt commented on modules/python/test/test_imread.py:
```diff
@@ -0,0 +1,30 @@
+from tests_common import NewOpenCVTests
+
+class dft_test(NewOpenCVTests):
+    def test_dft(self):
```
There is no `dtf` - rename the test

dkurt commented on modules/python/test/test_imread.py:
```diff
@@ -0,0 +1,30 @@
+        img = np.zeros((1000, 1000), dtype=np.uint8)
+        subImg = img[64:576, 128:640]
+        subImg = cv.imread(self.find_file("cv/shared/lena.png"), subImg, cv.IMREAD_GRAYSCALE)
+        ori = cv.imread(self.find_file("cv/shared/lena.png"), cv.IMREAD_GRAYSCALE)
```
Why grayscale?

dkurt commented on modules/python/test/test_imread.py:
```diff
@@ -0,0 +1,30 @@
+
+        img = np.zeros((1000, 1000), dtype=np.uint8)
+        subImg = img[64:576, 128:640]
+        subImg = cv.imread(self.find_file("cv/shared/lena.png"), subImg, cv.IMREAD_GRAYSCALE)
```
What's going on at this line?

dkurt commented on modules/python/test/test_imread.py:
```diff
@@ -0,0 +1,30 @@
+
+        img = np.zeros((1000, 1000), dtype=np.uint8)
+        subImg = img[64:576, 128:640]
+        subImg = cv.imread(self.find_file("cv/shared/lena.png"), subImg, cv.IMREAD_GRAYSCALE)
```
Can it work without assignment?
```python
cv.imread(self.find_file("cv/shared/lena.png"), subImg, cv.IMREAD_GRAYSCALE)
```

thewoz replied:
sorry i don't understand what you mean

dkurt replied:
Can this method work without an assignment?
replace
```python
subImg = cv.imread(self.find_file("cv/shared/lena.png"), subImg)
```
to
```python
cv.imread(self.find_file("cv/shared/lena.png"), subImg)
```

opencv-alalek commented:
> `OpenCV samples: Can't find required data file: cv/shared/lena.png in function 'findFile'`

You should not use API for "samples" in "tests".
Take a look on other tests.

thewoz commented:
Hi, @opencv-alalek I made the change. 
was that what you meant?

opencv-alalek commented on modules/imgcodecs/include/opencv2/imgcodecs.hpp:
```diff
@@ -227,6 +227,17 @@ Currently, the following file formats are supported:
+@note
+The image passing through the img parameter can be pre-allocated. The memory is reused if the shape and the type match with the load image.
+ */
+CV_EXPORTS_W void imread( const String& filename, OutputArray & dst, int flags = IMREAD_COLOR );
```
> `OutputArray &`

It is a double reference (see typedef). Remove one.

opencv-alalek commented on modules/imgcodecs/test/test_png.cpp:
```diff
@@ -7,6 +7,21 @@ namespace opencv_test { namespace {
+  EXPECT_EQ(0, countNonZero(img.rowRange(0, 1000).colRange(0, 128)));
+  EXPECT_EQ(0, countNonZero(img.rowRange(0, 1000).colRange(640, 1000)));
+  EXPECT_EQ(0, countNonZero(img.rowRange(0, 64).colRange(128, 640)));
+  EXPECT_EQ(0, countNonZero(img.rowRange(576, 1000).colRange(128, 640)));
```
`countNonZero()` => `norm(img..., NORM_INF)`

dkurt commented:
As discussed on team meeting, let's check Mat is contiguous.

opencv-alalek commented on modules/imgcodecs/src/loadsave.cpp:
```diff
@@ -457,7 +457,16 @@ imread_( const String& filename, int flags, Mat& mat )
+    else
+    {
+        CV_Assert(size == mat.size());
+        CV_Assert(type == mat.type());
```
```
CV_CheckEQ(size, mat.size(), "");
CV_CheckType(type, type == mat.type(), "");
```
